src/library/scala/concurrent/Future.scala

```diff
@@ -731,16 +732,24 @@ object Future {
     val i = futures.iterator
     if (!i.hasNext) Future.never
     else {
+      val deregs = ListBuffer.empty[() => Unit]
+      val l = new CountDownLatch(1)
       val p = Promise[T]()
       val firstCompleteHandler = new AtomicReference[Promise[T]](p) with (Try[T] => Unit) {
         override final def apply(v1: Try[T]): Unit =  {
           val r = getAndSet(null)
-          if (r ne null)
+          if (r ne null) {
             r tryComplete v1 // tryComplete is likely to be cheaper than complete
+            Future {
+              l.await()
+              deregs.foreach(_.apply())
```

Comment:
Is ListBuffer safe here?

Reply:
Yes, because the `l.await` and `deregs.foreach` are in a happens-before relationship, `l.countDown()` and `l.await()` are in a happens-before relationship, and `deregs.+=` and `l.countDown()` are in a happens-before relationship.
But, see my longer comment below.